### ISSUE TITLE
Certain challenges of YOLO11-OBB in document layout analysis scenarios

### ISSUE THREAD

**Issue by pearl88** [ultralytics/ultralytics#21069]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

First of all, thank you for creating such an outstanding project! When training on document data using YOLO11-OBB, I've achieved excellent precision with mAP50-95 reaching around 95. However, I'm encountering inaccurate bounding box localization issues for certain objects like text blocks, large tables, and images. Despite adjusting multiple training parameters, the improvements have been minimal.

I'm writing to inquire whether specific network architecture adjustments or configuration optimizations are recommended for document-centric scenarios in YOLO11-OBB. If such modifications exist, I would be deeply grateful for your expert guidance. Thank you immensely for your support!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @pearl88, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖 to help guide you more quickly! An Ultralytics engineer will also review and assist you soon.

**Comment by glenn-jocher** [comment]
Thanks for the detailed question! For document layout analysis with YOLO11-OBB, the localization issues you're experiencing with text blocks, tables, and images are common due to the unique aspect ratios and scales in document layouts.

Since you're achieving high mAP but poor localization, I'd recommend focusing on the box regression loss function - try experimenting with different IoU-based losses like `box=7.5` (default) vs higher values, and consider the `dfl=1.5` (Distribution Focal Loss) parameter which can improve localization accuracy for complex shapes.

For document-specific scenarios, also try increasing the `mosaic=0.0` to disable mosaic augmentation (which can hurt document layout learning), adjust `copy_paste=0.0`, and experiment with higher `iou=0.7` thresholds during training. The [hyperparameter tuning guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/) provides systematic approaches for optimizing these settings.

Document layouts often benefit from longer training schedules due to their complexity, so consider increasing epochs if you haven't already. What specific localization patterns are you seeing - are the boxes consistently offset in particular directions, or randomly scattered?

**Comment by pearl88** [comment]
@glenn-jocher 
 First and foremost, thank you so much for taking precious time from your busy schedule to answer my question and providing such detailed answers. Below are the detailed notes of my training process:

The dataset used is the open-source Publaynet, with 100k training images randomly selected.

Here is my args.yaml file; the main modifications include:

​Data Augmentation: Only random rotation is added, while all other augmentations are disabled (the original dataset does not support OBB; this configuration adjusts it to generate rotation-oriented data).
​box: Increased to ​15.

`task: obb
mode: train
model: yolo11l-obb.yaml
data: /ultralytics/ultralytics/cfg/datasets/publaynet_obb.yaml
epochs: 300
time: null
patience: 300
batch: 32
imgsz: 640
save: true
save_period: -1
cache: false
device: 0,1
workers: 8
project: null
name: train3
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 0
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 15
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 10.0
translate: 0.0
scale: 0.0
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.0
bgr: 0.0
mosaic: 0.0
mixup: 0.0
cutmix: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: randaugment
erasing: 0.4
cfg: null
tracker: botsort.yaml
save_dir: /ultralytics/runs/obb/train3`

Below are the batch images output by Ultralytics during training (via the plot_images function):

![Image](https://github.com/user-attachments/assets/073ca1d6-3700-403d-bb50-90bb005f2b0c)

Below is the training log output from my final epoch:
`300,290552,0.15149,0.07062,0,0.98372,0.9771,0.98708,0.96964,0.33814,0.20459,0,0.000133,0.000133,0.000133`

Regarding issues observed during inference testing:
For single-line, two-line, and three-line text detection blocks, ​the detection boxes are smaller than the actual text regions.

![Image](https://github.com/user-attachments/assets/8d6579b0-7c46-4e16-a75f-5f3521296f7e)

![Image](https://github.com/user-attachments/assets/c0b17d8b-b5a8-4530-90b9-21db45a5b2e6)

![Image](https://github.com/user-attachments/assets/d9e1a47c-3d99-4bbd-a062-08ca6c559d25)

![Image](https://github.com/user-attachments/assets/a962df4c-37c7-4e2b-8651-8881b8c1796c)

![Image](https://github.com/user-attachments/assets/f6272376-572c-4d52-ae71-1a778c36640e)

**Comment by pearl88** [comment]
@glenn-jocher 
Supplemental Training Result (I disabled dfl):​​
The box loss is still decreasing, suggesting 300 epochs might be insufficient. I will adjust to 500 epochs for testing.

![Image](https://github.com/user-attachments/assets/beab8a84-2edf-4bc3-b93d-3c0d0d6ec780)

**Comment by glenn-jocher** [comment]
Thanks for the comprehensive follow-up! Your configuration looks well-optimized for document analysis. The consistent pattern of boxes being smaller than actual text regions suggests a systematic bias in box regression rather than random localization errors.

Given that your box loss is still decreasing at epoch 300, extending to 500 epochs is definitely worthwhile. However, for the undersized box issue specifically, try these adjustments:

1. **Increase `dfl` back to 1.5 or even 2.0** - Distribution Focal Loss helps with precise localization and you may have removed a key component by disabling it
2. **Experiment with `warmup_bias_lr: 0.2`** (double the default) - this can help the model learn better initial box priors for document layouts
3. **Add slight `scale: 0.1`** augmentation - this helps the model learn to handle slight size variations which may improve box fitting

The undersized detection pattern often indicates the model is being overly conservative with box boundaries. Since Publaynet has challenging text block annotations, consider validating a few ground truth annotations manually to ensure they're tight around text boundaries.

Your current `box: 15` is quite aggressive - you might also try `box: 20` combined with `dfl: 2.0` to further emphasize accurate localization. The high mAP95 suggests your model has learned the features well, so these regression-focused adjustments should help with the box sizing.

**Comment by pearl88** [comment]
@glenn-jocher I'm truly delighted to have received your response! We've already made the parameter adjustments per your advice and launched a new training session. I'll keep you updated immediately once we have results. Once again, I greatly appreciate your generous guidance – your willingness to share expertise has been invaluable!

**Comment by glenn-jocher** [comment]
Excellent! I'm glad the recommendations resonate with your approach. Document layout analysis with OBB is a fascinating application and your systematic methodology of testing parameter adjustments is exactly the right approach.

Looking forward to hearing how the increased `dfl`, adjusted `warmup_bias_lr`, and slight scale augmentation perform with the extended 500 epoch training. The fact that you're seeing continued box loss improvement suggests there's still optimization headroom, which is promising for addressing those undersized detection patterns.

Feel free to share the results when ready - this type of real-world application feedback helps improve YOLO11-OBB for the broader community working on document AI tasks!

**Comment by pearl88** [comment]
@glenn-jocher First of all, I am very glad to have your attention and selfless assistance. After increasing DFL, adjusting warmup_bias_lr, and applying slight scale augmentation during an extended 500 training epochs, testing revealed that bounding box inaccuracies still persist. Typically, more than one problematic detection box exists per image. The best-performing model achieved so far is the one whose code was provided above, which exhibits a detection box issue only once every several images. Despite multiple rounds of parameter adjustments afterwards, no significant improvement has been observed. This situation leaves me quite perplexed, and I wonder whether structural optimizations to the model should be considered......

**Comment by glenn-jocher** [comment]
You've done excellent systematic parameter tuning, and since the undersized box issue persists despite comprehensive adjustments, it's time to consider structural approaches.

Given that Publaynet wasn't originally designed for OBB and you're generating rotated annotations, the ground truth box sizes might be inherently conservative. I'd recommend two structural approaches:

**1. Image Tiling Strategy:** For document layouts, try dividing your 640x640 images into overlapping tiles (e.g., 4x 320x320 with 25% overlap). This increases the effective resolution for text detection and often improves tight box fitting for small text elements - similar to the [image tiling approach for small objects](https://docs.ultralytics.com/guides/model-evaluation-insights/#image-tiling-for-small-objects).

**2. Custom Loss Weighting:** Since your box regression seems systematically biased, consider implementing a custom loss function that penalizes undersized boxes more heavily than oversized ones. You could modify the IoU calculation to asymmetrically weight box size errors.

The persistent nature suggests this might be a dataset annotation characteristic rather than a model limitation. Have you validated that your rotated OBB generation from the original Publaynet annotations maintains the tight text boundaries? Sometimes the conversion process can introduce systematic box shrinkage.

If the tiling approach doesn't help, we might need to explore custom architectural modifications or alternative training strategies specific to document layout characteristics.

**Comment by pearl88** [comment]
@glenn-jocher I am very pleased to receive your professional response once again. Regarding the accuracy of the Publaynet annotated data, we have conducted thorough checks ​both after format conversion and during training with data augmentation, and ​no annotation issues were found. Regarding your current suggestions, ​I will actively follow up and implement them. Thank you once again!

**Comment by glenn-jocher** [comment]
Perfect! Your meticulous approach to validation and systematic testing gives me confidence these structural approaches will yield meaningful insights. The thoroughness you've shown in both parameter tuning and annotation verification demonstrates excellent methodology for tackling complex document AI challenges.

Looking forward to seeing how the image tiling strategy performs with your document layouts - this could be a valuable contribution to the community's understanding of OBB optimization for document analysis scenarios.
